platform/android/export/export_plugin.cpp

```diff
@@ -525,6 +530,22 @@ bool EditorExportPlatformAndroid::is_package_name_valid(const String &p_package,
 		return false;
 	}
 
+	if (p_package.find("$genname") >= 0 && !is_project_name_valid(p_package)) {
+		if (r_error) {
+			*r_error = TTR("The end result may be unexpected. Try to explicitly specify the project name instead of the '$genname'.");
+		}
+		return false;
+	}
+
+	return true;
+}
+
+bool EditorExportPlatformAndroid::is_project_name_valid(const String &p_package) const {
+	String basename = GLOBAL_GET("application/config/name");
+	basename = basename.to_lower();
+	if (basename != get_valid_basename()) {
+		return false;
+	}
```

Comment:
Looks good! Can you add a comment describing this logic.  Otherwise it could be easy for a contributor to not realize that `basename` may not be equal to `get_valid_basename()` and mistakenly remove this method.